File: Java/328-odd-even-linked-list.java
// Given the head of a singly linked list, group all the nodes with odd indices together followed by the nodes with even indices, and return the reordered list.

// The first node is considered odd, and the second node is even, and so on.

// Note that the relative order inside both the even and odd groups should remain as it was in the input.

// You must solve the problem in O(1) extra space complexity and O(n) time complexity.

class Solution {
    public ListNode oddEvenList(ListNode head) {
        //base case
        if(head == null || head.next == null) return head;

        ListNode odd = head, even = head.next, evenHead = head.next;

        while(even != null && even.next != null) {
            odd.next = odd.next.next; //move pointer(next) to next odd node
            even.next = even.next.next; //move pointer(next) to next even node;
            odd = odd.next; //move odd to next odd node;
            even = even.next; //move even to next even node
        }
        odd.next = evenHead; //connect last odd node with first even node
        return head; //return head of the linked list;
    }  
}

File: Python/328_odd_even_linked_list.py
#Given the head of a singly linked list, group all the nodes with odd indices together followed by the nodes with even indices, and return the reordered list.
#The first node is considered odd, and the second node is even, and so on.
#Note that the relative order inside both the even and odd groups should remain as it was in the input.
#You must solve the problem in O(1) extra space complexity and O(n) time complexity.

#Example 1:
#Input: head = [1,2,3,4,5]
#Output: [1,3,5,2,4]

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head
        odd = head
        even = head.next
        even_head = head.next
        while even is not None and even.next is not None:
            odd.next = odd.next.next
            odd = odd.next
            even.next = even.next.next
            even = even.next
        odd.next = even_head
        return head
    